pkg/object/autocertmanager/autocertmanager.go

```diff
@@ -0,0 +1,417 @@
+		EnableHTTP01    bool         `yaml:"enableHTTP01"`
+		EnableTLSALPN01 bool         `yaml:"enableTLSALPN01"`
+		EnableDNS01     bool         `yaml:"enableDNS01"`
+		Domains         []DomainSpec `yaml:"domains" jsonschema:"required"`
+	}
+
+	// DomainSpec is the automate certificate management spec for a domain
+	DomainSpec struct {
+		Name        string            `yaml:"name" jsonschema:"required"`
+		DNSProvider map[string]string `yaml:"dnsProvider" jsonschema:"omitempty"`
+	}
+
+	// CertificateStatus is the certificate status of a domain
+	CertificateStatus struct {
+		Name       string    `yaml:"name"`
+		ExpireTime time.Time `yaml:"expireTime"`
+	}
+
+	// Status is the status of AutoCertManager
+	Status struct {
+		Domains []CertificateStatus `yaml:"domains"`
+	}
+)
+
+var (
+	globalACM atomic.Value
+)
+
+// Validate validates the spec of AutoCertManager
+func (spec *Spec) Validate() error {
+	if !(spec.EnableHTTP01 || spec.EnableTLSALPN01 || spec.EnableDNS01) {
+		return fmt.Errorf("at least one challenge type must be enabled")
+	}
+
+	for i := range spec.Domains {
+		d := &spec.Domains[i]
+
+		// convert to puny code to support Chinese or other unicode domain names
+		_, err := idna.Lookup.ToASCII(d.Name)
+		if err != nil && d.Name[0] == '*' {
+			_, err = idna.Lookup.ToASCII(d.Name[1:])
+		}
+		if err != nil {
+			return fmt.Errorf("domain name contains invalid characters: %s", d.Name)
+		}
+
+		if d.Name[0] != '*' {
+			continue
+		}
+
+		if !spec.EnableDNS01 {
+			return fmt.Errorf("find wildcard domain name but DNS-01 challenge is disabled: %s", d.Name)
+		}
+
+		if _, err := newDNSProvider(d); err != nil {
+			return fmt.Errorf("DNS provider configuration is invalid: %v", err)
+		}
+	}
+	return nil
+}
+
+// Zone returns the zone the domain belongs to
+func (spec *DomainSpec) Zone() string {
+	if spec.DNSProvider != nil {
+		return spec.DNSProvider["zone"]
+	}
+	return ""
+}
+
+func init() {
+	supervisor.Register(&AutoCertManager{})
+}
+
+// Category returns the category of AutoCertManager.
+func (acm *AutoCertManager) Category() supervisor.ObjectCategory {
+	return Category
+}
+
+// Kind return the kind of AutoCertManager.
+func (acm *AutoCertManager) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of AutoCertManager.
+func (acm *AutoCertManager) DefaultSpec() interface{} {
+	return &Spec{
+		DirectoryURL:    acme.LetsEncryptURL,
+		RenewBefore:     "720h",
+		EnableHTTP01:    true,
+		EnableTLSALPN01: true,
+		EnableDNS01:     true,
+	}
+}
+
+// Init initializes AutoCertManager.
+func (acm *AutoCertManager) Init(superSpec *supervisor.Spec) {
+	acm.superSpec = superSpec
+	acm.spec = superSpec.ObjectSpec().(*Spec)
+	acm.super = superSpec.Super()
+
+	// TODO: remove this check after converting AutoCertManager to a system controller
+	if globalACM.Load() != nil {
+		logger.Warnf("an AutoCertManager instance is already exist")
```

Comment:
Is here use panic better? Force custom to use `egctl object update` not `egctl object create` when `AutoCertManager` already created?

Reply:
this issue is tricky.
suppose we have `autocertmanager1` and want to create `autocertmanager2`, 
according to current logic, even if we `panic` here, the spec of `autocertmanager2` will be stored into Etcd,
because we don't guarantee object creation order during startup, when we do restart, it may successfully create `autocertmanager2` but panic when creating `autocertmanager1`.

as both behaviors are incorrect and we will change `AutoCertManager` to a system controller later, we can just keep this warning message for now.

Reply:
Make sense.... Another little question. I see that `globalACM` only store value, not store nil when close. Will this cause problem?

Reply:
per the doc of [atomic.Value](https://golang.google.cn/pkg/sync/atomic/#Value.Store), putting nil into it will result in a panic.

Reply:
updated, will clear the `globalACM` when there's no AutoCertManager instance now.